File: general_training_scripts/housing_regression.py
#!/usr/bin/env python3
import os
import numpy as np
import pandas as pd
import torch
from torch import nn
from torch.optim import Adam
from torch.utils.data import DataLoader, TensorDataset
from scipy.stats import gaussian_kde
import matplotlib.pyplot as plt

def plot_2d_kde(real_values, predicted_values, set_name="train", epoch=1, output_dir="training_output"):
    x, y = real_values.flatten(), predicted_values.flatten()
    kd = gaussian_kde(np.vstack([x, y]))
    xi = np.linspace(x.min(), x.max(), 256)
    yi = np.linspace(y.min(), y.max(), 256)
    X, Y = np.meshgrid(xi, yi)
    Z = kd(np.vstack([X.ravel(), Y.ravel()])).reshape(X.shape)
    plt.figure()
    plt.imshow(Z, origin="lower", aspect="auto", extent=[x.min(), x.max(), y.min(), y.max()], cmap="viridis")
    plt.title(f"Real vs. Predicted (2D KDE)\n({set_name} set, epoch {epoch})")
    plt.xlabel("Real Value")
    plt.ylabel("Predicted Value")
    os.makedirs(output_dir, exist_ok=True)
    fname = os.path.join(output_dir, f"2d_kde_{set_name}_epoch{epoch}.png")
    plt.savefig(fname)
    plt.close()

def plot_kde_distribution(data, set_name="train", epoch=1, output_dir="training_output", plot_title="KDE Plot", x_label="Values", overlay_normal=False):
    d = data.flatten()
    kd = gaussian_kde(d)
    x_vals = np.linspace(d.min(), d.max(), 250)
    pdf = kd(x_vals)
    plt.figure()
    plt.plot(x_vals, pdf, label="KDE", linewidth=2)
    plt.title(plot_title)
    plt.xlabel(x_label)
    plt.ylabel("Density")
    if overlay_normal:
        norm_pdf = lambda x: np.exp(-0.5*x**2)/np.sqrt(2*np.pi)
        plt.plot(x_vals, [norm_pdf(v) for v in x_vals], label="Standard Normal PDF", linewidth=2, linestyle="--")
    plt.legend()
    os.makedirs(output_dir, exist_ok=True)
    fname = os.path.join(output_dir, plot_title.replace(" ","_") + f"_epoch{epoch}.png")
    plt.savefig(fname)
    plt.close()

def evaluate_model(model, X, Y, set_name="train", epoch=1, output_dir="training_output"):
    model.eval()
    with torch.no_grad():
        preds = model(torch.from_numpy(X).float()).cpu().numpy().flatten()
    mse = np.mean((Y - preds)**2)
    rmse = np.sqrt(mse)
    residuals = preds - Y
    var_res = np.var(residuals) if len(residuals) > 1 else 0.0
    calibration_values = residuals / (np.sqrt(var_res) + 1e-8)
    within_ = lambda r: np.mean(np.abs(calibration_values) <= r)
    print(f"Percentage of calibration values within:")
    print(f"[-1, 1]: {within_(1)}")
    print(f"[-2, 2]: {within_(2)}")
    print(f"[-3, 3]: {within_(3)}")
    plot_kde_distribution(calibration_values, set_name, epoch, output_dir, f"Calibration vs Standard Normal\n({set_name} set, epoch {epoch})", "z = (μ̂ - y) / σ̂", True)
    plot_kde_distribution(residuals, set_name, epoch, output_dir, f"Residuals\n({set_name} set, epoch {epoch})", "residual = μ̂ - y", False)
    plot_kde_distribution(Y, set_name, epoch, output_dir, f"Real Values\n({set_name} set, epoch {epoch})", "Real Value", False)
    plot_kde_distribution(preds, set_name, epoch, output_dir, f"Predicted Values\n({set_name} set, epoch {epoch})", "Predicted Value", False)
    plot_2d_kde(Y, preds, set_name, epoch, output_dir)
    print(f"=== Evaluation on {set_name} set (epoch {epoch}) ===")
    print(f"MSE  = {mse}")
    print(f"RMSE = {rmse}\n")

def main():
    np.random.seed(98)
    torch.manual_seed(98)
    df = pd.read_csv("datasets/housing.csv")
    df = df.fillna(0)
    df["ocean_proximity"] = df["ocean_proximity"].replace({"ISLAND":"0.0","NEAR BAY":"1.0","NEAR OCEAN":"2.0","<1H OCEAN":"3.0","INLAND":"4.0"}).astype(float)
    y = df["median_house_value"].values
    X = df.drop("median_house_value", axis=1).values
    X = X.T
    Y = y.reshape(1, -1)
    X_train, X_val = X[:, :int(0.8*X.shape[1])], X[:, int(0.8*X.shape[1]):]
    Y_train, Y_val = Y[:, :int(0.8*Y.shape[1])], Y[:, int(0.8*Y.shape[1]):]
    X_mean, X_std = X_train.mean(axis=1, keepdims=True), X_train.std(axis=1, keepdims=True) + 1e-8
    Y_mean, Y_std = Y_train.mean(), Y_train.std() + 1e-8
    X_train = (X_train - X_mean)/X_std
    X_val = (X_val - X_mean)/X_std
    Y_train = (Y_train - Y_mean)/Y_std
    Y_val = (Y_val - Y_mean)/Y_std
    model = nn.Sequential(nn.Linear(X_train.shape[0],64), nn.ReLU(), nn.Linear(64,64), nn.ReLU(), nn.Linear(64,1))
    optimizer = Adam(model.parameters(), lr=1e-3)
    criterion = nn.MSELoss()
    batch_size = 64
    train_data = TensorDataset(torch.from_numpy(X_train.T).float(), torch.from_numpy(Y_train.T).float())
    train_loader = DataLoader(train_data, batch_size=batch_size, shuffle=True)
    num_epochs = 3
    num_training_its = 3
    print("Starting training on California Housing dataset...")
    train_losses = []
    validation_losses = []
    for epoch in range(1, num_epochs+1):
        model.train()
        for _ in range(num_training_its):
            for xb, yb in train_loader:
                optimizer.zero_grad()
                pred = model(xb).view(-1)
                loss = criterion(pred, yb.view(-1))
                loss.backward()
                optimizer.step()
                train_losses.append(loss.item())
                
                with torch.no_grad():
                    val_preds = model(torch.from_numpy(X_val.T).float()).cpu().numpy().flatten()
                    val_loss = criterion(torch.from_numpy(val_preds), torch.from_numpy(Y_val.flatten())).item()
                    validation_losses.append(val_loss)
                    
        evaluate_model(model, X_train.T, Y_train.flatten(), "train", epoch)
        evaluate_model(model, X_val.T, Y_val.flatten(), "val", epoch)
        print(f"Completed epoch {epoch}/{num_epochs}.")
    print("Training finished.")
    
    EMA_losses = [train_losses[0]]
    for i in range(1, len(train_losses)):
        EMA_losses.append(0.9*EMA_losses[-1] + 0.1*train_losses[i])
        
    EMA_losses_val = [validation_losses[0]]
    for i in range(1, len(validation_losses)):
        EMA_losses_val.append(0.9*EMA_losses_val[-1] + 0.1*validation_losses[i])
    
    # Plot line plot with training losses
    plt.figure()
    plt.plot(EMA_losses, label="Training Loss", linewidth=2)
    plt.plot(EMA_losses_val, label="Validation Loss", linewidth=2)
    plt.title("Training Loss")
    plt.xlabel("Iteration")
    plt.ylabel("Loss")
    plt.legend()
    plt.savefig("training_output/training_loss.png")
    plt.close()

if __name__ == "__main__":
    main()

File: general_training_scripts/housing_regression.jl
#!/usr/bin/env julia

# ============================================================================
# Example training script for California Housing Prices
# with Dataset struct + train/val split
# ============================================================================
# This version replaces the scatter plot with a 2D KDE plot.
# ============================================================================

# -- Load libraries and local code -------------------------------------------
include("../lib/utils.jl")        # Utility functions
include("../lib/datasets.jl")    # Contains the Dataset struct & constructor
include("../lib/factor_graph.jl") # Factor graph functionality

using Pkg
using CSV
using DataFrames
using Random

using Statistics
# using StatsBase          # for quantile
using KernelDensity      # for kde (supports 1D or 2D, etc.)
using Plots              # for plotting
using FilePathsBase: mkpath  # or Base.Filesystem.mkpath

# ---------------- 2D KDE plot function --------------------------------------
function plot_2d_kde(
    real_values,
    predicted_values;
    set_name = "train",
    epoch = 1,
    output_dir = "training_output"
)
    # Flatten to 1D vectors just in case they are matrices:
    x = vec(real_values)
    y = vec(predicted_values)

    # Compute 2D KDE
    kd = kde((x, y))

    # Create grid of points for plotting
    x_grid = collect(range(minimum(x), maximum(x), length=256))
    y_grid = collect(range(minimum(y), maximum(y), length=256))

    # Create a heatmap of the 2D density
    plt = heatmap(
        x_grid,          # Use collected grid instead of kd.x[1]
        y_grid,          # Use collected grid instead of kd.x[2]
        kd.density,
        title  = "Real vs. Predicted (2D KDE)\n($(set_name) set, epoch $epoch)",
        xlabel = "Real Value",
        ylabel = "Predicted Value",
        fill   = true,
        color  = :viridis
    )

    mkpath(output_dir)
    filename = joinpath(output_dir, "2d_kde_$(set_name)_epoch$(epoch).png")
    savefig(plt, filename)
    println("Saved 2D KDE plot to: $filename")
end

# ---------------- 1D KDE plot function --------------------------------------
function plot_kde_distribution(
    data;
    set_name = "train",
    epoch = 1,
    output_dir = "training_output",
    plot_title = "KDE Plot",
    x_label = "Values",
    y_label = "Density",
    overlay_normal = false
)
    kd = kde(vec(data))
    x_vals = range(minimum(kd.x), maximum(kd.x), length=250)

    # Optional standard normal PDF
    norm_pdf(x) = @inline (1 / sqrt(2π)) * exp(-0.5 * x^2)

    mkpath(output_dir)

    plt = plot(
        kd.x,
        kd.density,
        label    = "KDE",
        linewidth = 2,
        title    = plot_title,
        xlabel   = x_label,
        ylabel   = y_label
    )

    if overlay_normal
        plot!(
            x_vals,
            [norm_pdf(x) for x in x_vals],
            label     = "Standard Normal PDF",
            linewidth = 2,
            linestyle = :dash
        )
    end

    # For the filename, replace spaces in the title with underscores and append epoch
    filename = joinpath(output_dir, replace(plot_title, " " => "_") * "_epoch$(epoch).png")
    savefig(plt, filename)
    println("Saved KDE plot to: $filename")
end

"""
    evaluate_model(fg, X, Y; set_name="train", epoch=1, output_dir="training_output")

Evaluate the model by predicting on (X, Y), computing MSE, RMSE, variance percentiles,
and checking calibration by comparing (pred_mean - y) / pred_std to a standard
normal distribution. A kernel density plot of that distribution vs. the standard
normal is saved to `output_dir`. Also plots the residual distribution, real values,
predicted values, and a 2D KDE of (real, predicted).

# Arguments
- `fg::FactorGraph`: your trained (or in-training) factor graph
- `X::AbstractArray`: features
- `Y::AbstractArray`: ground truth targets
- `set_name::String`: e.g. "train" or "val"
- `epoch::Int`: current epoch, used for labeling output
- `output_dir::String`: folder to store plots (created if it doesn't exist)

"""
function evaluate_model(
    fg,
    X,
    Y;
    set_name = "train",
    epoch = 1,
    output_dir = "training_output"
)
    preds = predict(fg, X)
    pred_means = mean.(preds)
    pred_vars  = variance.(preds)

    mse  = Statistics.mean((Y .- pred_means).^2)
    rmse = sqrt(mse)

    ptiles = [0.10, 0.25, 0.50, 0.75, 0.90]
    var_percentiles = quantile(sqrt.(pred_vars), ptiles)

    calibration_values = (pred_means .- Y) ./ sqrt.(pred_vars)
    residuals = pred_means .- Y

    within_range(range) = count(abs.(calibration_values) .<= range) / length(calibration_values)
    println("Percentage of calibration values within:")
    println("[-1, 1]: ", within_range(1))
    println("[-2, 2]: ", within_range(2))
    println("[-3, 3]: ", within_range(3))

    # 1D KDE: calibration vs normal
    plot_kde_distribution(
        calibration_values;
        set_name = set_name,
        epoch = epoch,
        output_dir = output_dir,
        plot_title = "Calibration vs Standard Normal\n($(set_name) set, epoch $epoch)",
        x_label = "z = (μ̂ - y) / σ̂",
        overlay_normal = true
    )

    # 1D KDE: residual distribution
    plot_kde_distribution(
        residuals;
        set_name = set_name,
        epoch = epoch,
        output_dir = output_dir,
        plot_title = "Residuals\n($(set_name) set, epoch $epoch)",
        x_label = "residual = μ̂ - y",
        overlay_normal = false
    )

    # 1D KDE: real values
    plot_kde_distribution(
        Y;
        set_name = set_name,
        epoch = epoch,
        output_dir = output_dir,
        plot_title = "Real Values\n($(set_name) set, epoch $epoch)",
        x_label = "Real Value",
        overlay_normal = false
    )

    # 1D KDE: predicted values
    plot_kde_distribution(
        pred_means;
        set_name = set_name,
        epoch = epoch,
        output_dir = output_dir,
        plot_title = "Predicted Values\n($(set_name) set, epoch $epoch)",
        x_label = "Predicted Value",
        overlay_normal = false
    )

    # 2D KDE for Real vs. Predicted
    plot_2d_kde(Y, pred_means; set_name=set_name, epoch=epoch, output_dir=output_dir)

    println("=== Evaluation on $(set_name) set (epoch $epoch) ===")
    println("MSE  = $mse")
    println("RMSE = $rmse")
    println("Predicted variance percentiles at $(ptiles) => $var_percentiles\n")
end

# -- Set random seed for reproducibility -------------------------------------
Random.seed!(98)

# -- Load the housing dataset ------------------------------------------------
df = CSV.read("datasets/housing.csv", DataFrame)

total_missing = sum(ismissing.(Matrix(df)))
println("Total missing values: $total_missing")

# Replace missing values with 0 (column-by-column)
for c in names(df)
    replace!(df[!, c], missing => 0)
end
disallowmissing!(df)

# -- Replace ocean_proximity categories with float values --------------------
replace!(df."ocean_proximity", 
         "ISLAND"    => "0.0",
         "NEAR BAY"  => "1.0",
         "NEAR OCEAN"=> "2.0",
         "<1H OCEAN" => "3.0",
         "INLAND"    => "4.0")
df."ocean_proximity" = parse.(Float64, df."ocean_proximity")

# -- Extract features (X) and target (Y) --------------------------------------
# Adjust the column name for the target as needed; here we assume
# the column for the house value is named :median_house_value

y = df."median_house_value"                             # Vector
X = select(df, Not(:median_house_value)) |> Matrix      # Matrix

# Typically factor graphs in the provided codebase expect data in the format
# (features..., batch). We'll transpose so shape = (#features, #samples).
X = X'               # shape: (num_features, num_samples)
Y = reshape(y, 1, :) # shape: (1, num_samples)

# -- Build our dataset with train/val split ----------------------------------
# The Dataset constructor (from datasets.jl) will split into X_train, Y_train, X_val, Y_val
# By default it uses 80% of the data for training, 20% for validation.
X = hcat(X)

println("Type of X: ", typeof(X))
println("Size of X: ", size(X))
println("Type of Y: ", typeof(Y))
println("Size of Y: ", size(Y))

dset = Dataset(X, Y)  # You can pass train_perc=0.8 if you want to customize
dset = normalize_X!(dset)
dset = normalize_Y!(dset)

println("Training set has size: ", size(dset.X_train), " => X_train")
println("Training set has size: ", size(dset.Y_train), " => Y_train")
println("Validation set has size: ", size(dset.X_val), " => X_val")
println("Validation set has size: ", size(dset.Y_val), " => Y_val")

# -- Generate desired model architecture -------------------------------------
function generate_architecture(input_shape, action_space_dimensions, regression_basic_architecture)
    return [
        (input_shape,),
        (:Linear, 64),
        (:LeakyReLU, 0.1),
        (:Linear, 64),
        (:LeakyReLU, 0.1),
        (:Linear, action_space_dimensions),
        (:Regression, regression_basic_architecture^2),
    ]
end

# For a simple single-output regression, set action_space_dimensions = 1
# and choose a small integer for regression_basic_architecture (like 1)
input_shape               = size(dset.X_train, 1)  # number of features
action_space_dimensions   = 1
regression_basic_arch_val = 0.4

architecture = generate_architecture(input_shape,
                                     action_space_dimensions,
                                     regression_basic_arch_val)

# -- Create the factor graph model -------------------------------------------
batch_size = 64
fg = create_factor_graph(architecture, batch_size)

# -- Create Trainer and train -----------------------------------------------
trainer = Trainer(fg, dset.X_train, dset.Y_train)

num_epochs       = 10
num_training_its = 3

training_losses = Vector{Float64}()
validation_losses = Vector{Float64}()

println("Starting training on California Housing dataset...")
train(trainer; num_epochs=num_epochs, num_training_its=num_training_its, silent=false, training_losses=training_losses, validation_losses=validation_losses, validation_X=dset.X_val, validation_Y=dset.Y_val)

# Evaluate on training data
evaluate_model(fg, dset.X_train, dset.Y_train; set_name="train", epoch=0)

# Evaluate on validation data
evaluate_model(fg, dset.X_val, dset.Y_val; set_name="val", epoch=0)
println("Training finished.")

EMA_losses = [training_losses[1]]
for i in 2:length(training_losses)
    push!(EMA_losses, 0.9 * EMA_losses[end] + 0.1 * training_losses[i])
end

EMA_val_losses = [validation_losses[1]]
for i in 2:length(validation_losses)
    push!(EMA_val_losses, 0.9 * EMA_val_losses[end] + 0.1 * validation_losses[i])
end

# Clamp EMA losses to [0, 2]
EMA_losses = clamp.(EMA_losses, 0, 2)
EMA_val_losses = clamp.(EMA_val_losses, 0, 2)

# Create line plot for EMA losses.
plot(EMA_losses, label="EMA Train Loss", xlabel="Iteration", ylabel="Loss", title="Training vs Validation EMA Loss")
plot!(EMA_val_losses, label="EMA Val Loss")
savefig("training_output/ema_loss_plot.png")
